Deprecations to create in 2.6.0

weierophinney opened the pull request:
Per #4, we need to mark a number of features as deprecated for an upcoming 2.6.0 release. This should be done against master.

Deprecations can be done via annotations; all deprecations **MUST** also be noted in the CHANGELOG.md.
- [x] Mark GlobalEventManager as deprecated
- [x] Mark StaticEventManager as deprecated
- [x] Mark EventManager(Interface)?::setSharedManager as deprecated
- [x] Mark EventManager(Interface)?::getEvents as deprecated
- [x] Mark EventManager(Interface)?::getListeners as deprecated
- [x] Mark EventManager(Interface)?::setEventClass as deprecated
- [x] Mark EventManager(Interface)?::attachAggregate as deprecated
- [x] Mark EventManager(Interface)?::detachAggregate as deprecated
- [x] Mark SharedListenerAggregateInterface as deprecated
- [x] Mark SharedEventAggregateAwareInterface as deprecated
- [x] Mark SharedEventManagerAwareInterface as deprecated
- [x] Mark SharedEventManagerInterface::getEvents as deprecated
- [x] Add EventManager::triggerEvent() (but not at interface level)
- [x] Add EventManager::triggerEventUntil() (but not at interface level)

gianarb commented:
I can tag this methods but can you confirm me this way?

``` php
/**
 * @deprecated This method will be removed since version 3.0
 */
```

``` php
/**
 * @deprecated This class will be removed since version 3.0
 */
```

Thanks

weierophinney commented:
@gianarb Let's go with:

``` php
/**
 * @deprecated This {class or method} is deprecated with 2.6.0, and will be removed in 3.0.0. Please see {link to doc/book/migration/removed.md} for details.
 */
```

I need to figure out where the link will live; likely the develop branch, however.

weierophinney commented:
The link will be https://github.com/zendframework/zend-eventmanager/blob/develop/doc/book/migration/removed.md

@gianarb Ping me on IRC if you see this; I plan to start work on it unless I hear from you shortly. :)

weierophinney committed:
Deprecations/additions for 2.6.0

- [X] Mark GlobalEventManager as deprecated
- [X] Mark StaticEventManager as deprecated
- [X] Mark EventManager::setSharedManager as deprecated
  - [X] Add SharedEventsCapableInterface
  - [X] Updated SharedEventManagerAwareInterface to extend SharedEventsCapableInterface
- [X] Mark EventManager::unsetSharedManager as deprecated
- [X] Mark EventManager(Interface)?::getEvents as deprecated
- [X] Mark EventManager(Interface)?::getListeners as deprecated
- [X] Mark EventManager(Interface)?::setEventClass as deprecated
- [X] Mark EventManager(Interface)?::attachAggregate as deprecated
- [X] Mark EventManager(Interface)?::detachAggregate as deprecated
- [X] Mark SharedListenerAggregateInterface as deprecated
- [X] Mark SharedEventAggregateAwareInterface as deprecated
- [X] Mark SharedEventManagerAwareInterface as deprecated
- [X] Mark SharedEventManager(Interface)?::getEvents as deprecated
- [X] Add EventManager::triggerEvent() (but not at interface level)
- [X] Add EventManager::triggerEventUntil() (but not at interface level)

akrabat commented on src/SharedEventManagerAwareInterface.php:
```diff
@@ -23,13 +27,6 @@
-     */
-    public function getSharedManager();
-
-    /**
```
This is a BC break, surely?

weierophinney committed:
Incorporate feedback

- Per @Akrabat, the "3.0.0" verbiage needed to be on the same line as the
  annotation.

weierophinney commented:
Discussed triggering `E_USER_DEPRECATED` with @akrabat, and decided **not** to do so, as it prematurely breaks a lot of code. (I observed this elsewhere in Apigility when ZF components triggered such errors, causing API payloads to break when `display_errors` is on — which is often useful in development!)

akrabat commented:
Looks good to me.
